rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,669 @@
+semantics to a `NonNull` raw pointer. This seems necessary to get reasonable
+ergonomics with an atomics-like wrapper type approach. However, it could also be
+argued that a `VolatileXyz::store(ptr, data, ordering)` style of API would work
+well enough, and avoid coupling with unstable features. Similarly, the use of
+`NonNull` itself could be debated. This will be revisited in the alternatives
+section of the RFC as well.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Deprecating the existing volatile operations will cause language churn. Not
+deprecating them will keep around two different and subtly incompatible ways to
+do the same thing, which is equally bad. It could be argued that the issues with
+the existing volatile operations, while real, do not warrant the full complexity
+of this RFC.
+
+As mentioned above, exposing more than loads and stores, and non-`Relaxed`
+atomic memory orderings, also muddies the "a volatile op should compile into one
+hardware memory instruction" narrative that is so convenient for loads and
+stores. Further, compatibility of non-load/store atomics in IPC scenario may
+require some ABI agreement on how atomics should be implemented between the
+interacting programs.
+
+If this is perceived to be a problem, we could decide to do away with some of
+the complexity by initially focusing on a restricted subset of this proposal
+which only supports `Relaxed` loads and stores, and saving more complex atomic
+operations as a future extension.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This effort seems worthwhile because it simultaneously eliminates two well-known
+and unnecessary volatile footguns (data races and tearing) and opens interesting
+new possibilities in interprocess communication.
+
+But although this feature may seem simple, its design space is actually
+remarquably large, and a great many alternatives were considered before reaching
+the current design proposal. Here are some design knobs that were explored:
+
+## Extending `AtomicXyz`
+[extending-atomicxyz]: #extending-atomicxyz
+
+Atomic volatile operations could be made part of the Atomic wrapper types.
+However, they cannot be just another atomic memory ordering, due to the fact
+that atomic ops take an `&self` (which is against the spirit of volatile as it
+is tagged with `dereferenceable` at the LLVM layer).
+
+Instead, one would need to add more methods to atomic wrapper types, which take
+a pointer as a self-parameter. This API inconsistency was felt to be
+unnecessarily jarring to users, not to mention that potentially having a copy of
+each atomic memory operation with a `_volatile` prefix would reduce the
+readability of the `AtomicXyz` types' API documentation.
+
+In addition, it is extremely common to want either all operations on a memory
+location to be volatile, or none of them. Providing separate wrapper types
+helps enforce this very common usage pattern at the API level.
```

Comment:
This is a good argument. OTOH, one could leave providing such safe wrappers to higher-level libraries. With the API proposed by the RFC, in the uncommon case of *wanting* to mix volatile and non-volatile (atomic) operations on the same locations, people need to resort to `transmute`.

Reply:
As a minor nitpick, if use of `transmute` were the only issue, one could just add a `VolatileXyz` method that casts to `AtomicXyz` with a clear warning that doing so voids your "absence of out-of-thin-air-reads" warranty.

Reply:
That is a good argument.

Reply:
Integrated in "Extending AtomicXyz" discussion.